### DISCUSSION TITLE
Starter Kit and Package elements

### DISCUSSION THREAD

**Discussion by woodseowl** [CornellCustomDev/CD-LaravelStarterKit#15] [General]
I understand there to be three kind of elements we build for our Laravel Starter Kit concept:

1. A client project based off the Starter Kit
2. The Starter Kit itself, implemented as a composer-included package with management features
3. Packages that we develop that can be included in the Starter Kit (default / mandatory) and packages that can optionally be included in a project

I think we know generally what the first and third of those generally look like, since we create Laravel projects and use third-party packages regularly. That said, we do not have much internal expertise with building Laravel packages (see the [Laravel Package Development](https://laravelpackage.com/) site).

I've been wondering what capabilities the Starter Kit should provide that wouldn't be encapsulated in a package. Examples:
- Standard files that we have like `.lando.yml` or other customized files like `.gitignore` or `.env.example` which we would place or update via an install script
- I saw that the the [Spatie package skeleton](https://github.com/spatie/package-skeleton-laravel) has a script that replaces placeholder names after you clone the template
- Standard GitHub actions, testing setups, code styling tools, and other things that are more about standard process than code

Here are some things I think we'd want in any Starter Kit implementation but that could be created as packages:
- Base layouts for Cornell branding and Custom Dev CSS framework
- Shibboleth / SAML login (incl. Cornell, Weill, and local user auth + management)

These are slightly less certain, but seem like they would be needed for most projects (possibly provided as third-party packages):
- LDAP / AD helpers
- Admin interface for models + configuration
- Navigation management (menus, page titles)

Then there are things that are not for every project but probably wouldn't be third-party packages. Examples:
- PeopleSoft / Canvas / Salesforce integrations

I'm also curious how we would want to track third-party packages that we want to be standard or recommended:
- Excel/CSV exporters and importers
- Lightweight CMS
- Front-end data tables (beyond basic styling from our CSS framework)

I looked through the [feature brainstorm Trello board](https://trello.com/b/hEhbtH0v/laravel-starter-kit-features-brainstorm) to see if there were other ideas that stood out. Everything I see looks kind of like one of the things above. I'm interested in what other people see. Please add to this!

**Comment by richmarisa** [comment]
I think it would be best to add things like the LDAP/AD helpers, admin interface and navigation (which may themselves be implemented as separate packages) to the base Starter Kit. As a comparison, there are lots of things in Laravel that I don't always use, but I expect they are there, and well integrated with the whole.    I'd also include a basic data table implementation -- a version of the one from GSC, which I copied into Find Facilities, would be a good candidate. 

I think we need to maintain a list of "recommended" third party packages and test compatibility before releasing new versions of the Starter Kit so we don't leave those packages behind.

**Reply by inaydich** [reply]
I agree with Rich the starter kit should lean and light. I do not think it needs Weill authentication or any integrations like PS or SF. I would include those extra functionalities into a demo site. Also CC Framework needs a good revision what files are really needed. I did a round of that in GSC, but maybe it is useful to do one more review and an update. 
At the same I think Livewire and lando config files are definitely should be there because we are using them everywhere.

**Comment by woodseowl** [comment]
I've created PR #16 as an attempt at building out some initial setup and addressing the first bullet point about starter kit aspects that wouldn't be in separate packages. I used the laravelpackage.com site to walk through creating it, so it largely follows their approach. I'd love to hear if what is in that PR is going the right direction and use it to help us frame out what we should build and where.
